lib/commands/preload.ts

```diff
@@ -351,43 +345,47 @@ Can be repeated to add multiple certificates.\
 		},
 	};
 
-	allDeviceTypes: DeviceTypeJson.DeviceType[];
-	async getDeviceTypes() {
-		if (this.allDeviceTypes === undefined) {
-			const balena = getBalenaSdk();
-			const deviceTypes = await balena.models.config.getDeviceTypes();
-			this.allDeviceTypes = _.sortBy(deviceTypes, 'name');
-		}
-		return this.allDeviceTypes;
-	}
-
 	isCurrentCommit(commit: string) {
 		return commit === 'latest' || commit === 'current';
 	}
 
-	async getDeviceTypesWithSameArch(deviceTypeSlug: string) {
-		const deviceTypes = await this.getDeviceTypes();
-		const deviceType = _.find(deviceTypes, { slug: deviceTypeSlug });
-		if (!deviceType) {
-			throw new Error(`Device type "${deviceTypeSlug}" not found in API query`);
-		}
-		return _(deviceTypes).filter({ arch: deviceType.arch }).map('slug').value();
-	}
-
 	async getApplicationsWithSuccessfulBuilds(deviceTypeSlug: string) {
 		const balena = getBalenaSdk();
 
-		const deviceTypes = await this.getDeviceTypesWithSameArch(deviceTypeSlug);
+		try {
+			await balena.models.deviceType.get(deviceTypeSlug);
+		} catch {
+			throw new Error(`Device type "${deviceTypeSlug}" not found in API query`);
+		}
 		return (await balena.models.application.getAll({
 			$select: ['id', 'app_name', 'should_track_latest_release'],
 			$expand: this.applicationExpandOptions,
 			$filter: {
+				// get the apps that are of the same arch as the device type of the image
 				is_for__device_type: {
 					$any: {
 						$alias: 'dt',
 						$expr: {
 							dt: {
-								slug: { $in: deviceTypes },
+								is_of__cpu_architecture: {
+									$any: {
+										$alias: 'ioca',
+										$expr: {
+											ioca: {
+												is_supported_by__device_type: {
+													$any: {
+														$alias: 'isbdt',
+														$expr: {
+															isbdt: {
+																slug: deviceTypeSlug,
```

Comment:
Would this work with device type aliases, e.g. where `deviceTypeSlug` was `nuc` instead of `intel-nuc`?
If not or in doubt, earlier the code was already executing:
```ts
    await balena.models.deviceType.get(deviceTypeSlug);
```
So I guess we could use the returned device type object to get its "canonical" slug (assuming that `models.deviceType.get` accepts device type aliases).

Reply:
Actually I am not sure if `deviceTypeSlug` could be an alias or not. It looks like the slug could come from the `config.json` file found in the OS image being preloaded:
* https://github.com/balena-io/balena-cli/blob/v12.48.5/lib/commands/preload.ts#L531
* https://github.com/balena-io-modules/balena-preload/blob/v10.5.0/lib/preload.js#L481
* https://github.com/balena-io-modules/balena-preload/blob/v10.5.0/src/preload.py#L929
* https://github.com/balena-io-modules/balena-preload/blob/v10.5.0/src/preload.py#L700

I guess someone could manually edit the `config.json` file and replace, say, "raspberrypi3" with "raspberrypi2" when using an old `config.json` file as a template, not realising that "raspberrypi2" is an alias for "raspberry-pi2".

By the way, the `balena-preload` repo also contains an occurrence of the deprecated `models.config.getDeviceTypes()`:
https://github.com/balena-io-modules/balena-preload/blob/v10.5.0/lib/preload.js#L309

cc: @klutchell as maintainer of `balena-preload` -- As "warned" by Alex M during the all hands call, we should remove usage of the API's `/device-types/v1`, which is used by the balena SDK's deprecated `models.config.getDeviceTypes()` method. Unless Thodoris says he will create a PR for `balena-preload` as well. :-)

Reply:
> Would this work with device type aliases

@pdcastro no it will not, but the previous implementation was also not working with aliases, so there is no change of behavior.

> earlier the code was already executing `await balena.models.deviceType.get(deviceTypeSlug);`

I don't see that in the diff. I only see previously doing `await balena.models.config.getDeviceTypes()`` and then filtering on the slug, which would not allow using aliases, since it's filtering on and then returning only the canonical DT slug.

Reply:
> By the way, the balena-preload repo also contains an occurrence of the deprecated models.config.getDeviceTypes():

Good catch, I already opened an issue for our usage of `deployArtifact` but I will rename it to cover any `/device-types/v1` endpoints.

Reply:
> It looks like the slug could come from the config.json file found in the OS image being preloaded:

This looks correct, we read it from `config.json` on the device and that should never be an alias. If someone used an alias there manually I would expect a number of issues either way.

Reply:
> earlier the code was already executing `await balena.models.deviceType.get(deviceTypeSlug);`

By "earlier" I meant a few lines of code before the line my comment was added to. Code introduced by this PR, rather than pre-existing code. `await balena.models.deviceType.get(deviceTypeSlug);` was introduced by this PR but the return value is discarded. If we didn't discard the return value, I thought we might take the "canonical" (non alias) slug from the return value, assuming that `models.deviceType.get(deviceTypeSlug)` works with an alias.

Reply:
> assuming that `models.deviceType.get(deviceTypeSlug)` works with an alias.

Well, it doesn't, I've just tested. `models.deviceType.get('nuc')` throws an error.

Reply:
No it doesn't since aliases are not yet part of the model.
Atm `models.deviceType.get(deviceTypeSlug)` is there just to throw a DeviceTypeNotFoundError, essentially replacing the `throw new Error(`Device type "${deviceTypeSlug}" not found in API query`);` that we had before.